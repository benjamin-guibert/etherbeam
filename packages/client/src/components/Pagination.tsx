import React, { ReactElement } from 'react'
import PaginationItem from './PaginationItem'
import './Pagination.scss'

interface PaginationProps {
  total: number
  current: number
  setCurrent: (next: number | ((previous: number) => number)) => void
  className?: string
}

const ELLIPSIS_THRESHOLD_PAGES = 7

const Pagination = ({ total, current, setCurrent, className }: PaginationProps): ReactElement => {
  const displayEllipsis = total > ELLIPSIS_THRESHOLD_PAGES
  const pages = []
  Array.from({ length: total }).map((_, index) => {
    const page = index + 1
    const isCurrent = page === current
    if (!displayEllipsis || [1, current - 1, current, current + 1, total].includes(page)) {
      pages.push(() => (
        <PaginationItem page={page} onClick={() => setCurrent(page)} selected={isCurrent} disabled={isCurrent} />
      ))
    } else if (page == current - 2 || page == current + 2) {
      pages.push(() => <PaginationItem page="ellipsis" disabled />)
    }
  })

  return (
    <nav>
      <ul className={['my-pagination my-hlist', className].join(' ')}>
        <PaginationItem page="first" onClick={() => setCurrent(1)} disabled={current === 1} />
        <PaginationItem
          page="previous"
          onClick={() => setCurrent((previous) => previous - 1)}
          disabled={current === 1}
        />
        {pages.map((Page, index) => (
          <Page key={index} />
        ))}
        <PaginationItem
          page="next"
          onClick={() => setCurrent((previous) => previous + 1)}
          disabled={current === total}
        />
        <PaginationItem page="last" onClick={() => setCurrent(total)} disabled={current === total} />
      </ul>
    </nav>
  )
}

export default Pagination
