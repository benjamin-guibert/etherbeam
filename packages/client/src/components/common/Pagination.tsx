import React, { ReactElement } from 'react'
import BsPagination from 'react-bootstrap/Pagination'

interface PaginationProps {
  total: number
  current: number
  setCurrent: (next: number | ((previous: number) => number)) => void
  size?: 'sm' | 'lg'
  className?: string
}

const ELLIPSIS_THRESHOLD_PAGES = 7

const Pagination = ({ total, current, setCurrent, className, size }: PaginationProps): ReactElement => {
  const displayEllipsis = total > ELLIPSIS_THRESHOLD_PAGES
  const pages = []
  Array.from({ length: total }).map((_, index) => {
    const page = index + 1
    if (!displayEllipsis || [1, current - 1, current, current + 1, total].includes(page)) {
      pages.push(() => (
        <BsPagination.Item key={page} active={page === current} onClick={() => setCurrent(page)}>
          {page}
        </BsPagination.Item>
      ))
    } else if (page == current - 2 || page == current + 2) {
      pages.push(() => <BsPagination.Ellipsis disabled />)
    }
  })

  return (
    <BsPagination className={className} size={size}>
      <BsPagination.First disabled={current === 1} onClick={() => setCurrent(1)} />
      <BsPagination.Prev disabled={current === 1} onClick={() => setCurrent((previous) => previous - 1)} />
      {pages.map((Page, index) => (
        <Page key={index} />
      ))}
      <BsPagination.Next disabled={current === total} onClick={() => setCurrent((previous) => previous + 1)} />
      <BsPagination.Last disabled={current === total} onClick={() => setCurrent(total)} />
    </BsPagination>
  )
}

export default Pagination
