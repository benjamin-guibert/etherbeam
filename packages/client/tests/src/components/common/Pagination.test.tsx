import * as React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import Pagination from 'components/common/Pagination'
import { act } from 'react-dom/test-utils'

describe('<Pagination />', () => {
  const setCurrent = jest.fn()

  let component: ReactWrapper

  describe('All', () => {
    beforeAll(() => {
      component = mount(<Pagination total={6} current={5} setCurrent={setCurrent} size="sm" className="app-class" />)
    })

    it('should render', () => {
      const pages = component.find('PageItem')
      expect(pages).toHaveLength(10)
      expect(pages.at(0).prop('disabled')).toBeFalsy()
      expect(pages.at(1).prop('disabled')).toBeFalsy()
      expect(pages.at(2).prop('active')).toBeFalsy()
      expect(pages.at(3).prop('active')).toBeFalsy()
      expect(pages.at(4).prop('active')).toBeFalsy()
      expect(pages.at(5).prop('active')).toBeFalsy()
      expect(pages.at(6).prop('active')).toBeTruthy()
      expect(pages.at(7).prop('active')).toBeFalsy()
      expect(pages.at(8).prop('disabled')).toBeFalsy()
      expect(pages.at(9).prop('disabled')).toBeFalsy()
    })
  })

  describe('Ellipsed', () => {
    beforeAll(() => {
      component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} size="sm" className="app-class" />)
    })

    it('should render pagination', () => {
      const pagination = component.find('Pagination')
      expect(pagination.prop('size')).toBe('sm')
      expect(pagination.prop('className')).toBe('app-class')
    })

    it('should render pages', () => {
      const pages = component.find('PageItem')
      expect(pages).toHaveLength(11)
      expect(pages.at(0).prop('disabled')).toBeFalsy()
      expect(pages.at(1).prop('disabled')).toBeFalsy()
      expect(pages.at(2).prop('active')).toBeFalsy()
      expect(pages.at(3).prop('disabled')).toBeTruthy()
      expect(pages.at(4).prop('active')).toBeFalsy()
      expect(pages.at(5).prop('active')).toBeTruthy()
      expect(pages.at(6).prop('active')).toBeFalsy()
      expect(pages.at(7).prop('disabled')).toBeTruthy()
      expect(pages.at(8).prop('active')).toBeFalsy()
      expect(pages.at(9).prop('disabled')).toBeFalsy()
      expect(pages.at(10).prop('disabled')).toBeFalsy()
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('First page', () => {
    beforeAll(() => {
      jest.resetAllMocks()

      component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)
      setCurrent.mockImplementation((newValue) => {
        component.setProps({ current: newValue })
      })

      act(() => {
        component.find('PageItem').at(0).find('a').simulate('click')
        component.update()
      })
    })

    it('should set current', () => {
      expect(setCurrent).toBeCalledTimes(1)
      expect(component.prop('current')).toBe(1)
    })

    it('should render pages', () => {
      const pages = component.find('PageItem')
      expect(pages).toHaveLength(8)
      expect(pages.at(0).prop('disabled')).toBeTruthy()
      expect(pages.at(1).prop('disabled')).toBeTruthy()
      expect(pages.at(2).prop('active')).toBeTruthy()
      expect(pages.at(3).prop('active')).toBeFalsy()
      expect(pages.at(4).prop('disabled')).toBeTruthy()
      expect(pages.at(5).prop('active')).toBeFalsy()
      expect(pages.at(6).prop('disabled')).toBeFalsy()
      expect(pages.at(7).prop('disabled')).toBeFalsy()
    })
  })

  describe('Previous page', () => {
    beforeAll(() => {
      jest.resetAllMocks()

      component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)
      setCurrent.mockImplementation((newValue) => {
        component.setProps({ current: newValue(5) })
      })

      act(() => {
        component.find('PageItem').at(1).find('a').simulate('click')
        component.update()
      })
    })

    it('should set current', () => {
      expect(setCurrent).toBeCalledTimes(1)
      expect(component.prop('current')).toBe(4)
    })

    it('should render pages', () => {
      const pages = component.find('PageItem')
      expect(pages).toHaveLength(11)
      expect(pages.at(0).prop('disabled')).toBeFalsy()
      expect(pages.at(1).prop('disabled')).toBeFalsy()
      expect(pages.at(2).prop('active')).toBeFalsy()
      expect(pages.at(3).prop('disabled')).toBeTruthy()
      expect(pages.at(4).prop('active')).toBeFalsy()
      expect(pages.at(5).prop('active')).toBeTruthy()
      expect(pages.at(6).prop('active')).toBeFalsy()
      expect(pages.at(7).prop('disabled')).toBeTruthy()
      expect(pages.at(8).prop('active')).toBeFalsy()
      expect(pages.at(9).prop('disabled')).toBeFalsy()
      expect(pages.at(10).prop('disabled')).toBeFalsy()
    })
  })

  describe('Specific page', () => {
    beforeAll(() => {
      jest.resetAllMocks()

      component = mount(<Pagination total={10} current={3} setCurrent={setCurrent} />)
      setCurrent.mockImplementation((newValue) => {
        component.setProps({ current: newValue })
      })

      act(() => {
        component.find('PageItem').at(3).find('a').simulate('click')
        component.update()
      })
    })

    it('should set current', () => {
      expect(setCurrent).toBeCalledTimes(1)
      expect(component.prop('current')).toBe(2)
    })

    it('should render pages', () => {
      const pages = component.find('PageItem')
      expect(pages).toHaveLength(9)
      expect(pages.at(0).prop('disabled')).toBeFalsy()
      expect(pages.at(1).prop('disabled')).toBeFalsy()
      expect(pages.at(2).prop('active')).toBeFalsy()
      expect(pages.at(3).prop('active')).toBeTruthy()
      expect(pages.at(4).prop('active')).toBeFalsy()
      expect(pages.at(5).prop('disabled')).toBeTruthy()
      expect(pages.at(6).prop('active')).toBeFalsy()
      expect(pages.at(7).prop('disabled')).toBeFalsy()
      expect(pages.at(8).prop('disabled')).toBeFalsy()
    })
  })

  describe('Next page', () => {
    beforeAll(() => {
      jest.resetAllMocks()

      component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)
      setCurrent.mockImplementation((newValue) => {
        component.setProps({ current: newValue(5) })
      })

      act(() => {
        component.find('PageItem').at(9).find('a').simulate('click')
        component.update()
      })
    })

    it('should set current', () => {
      expect(setCurrent).toBeCalledTimes(1)
      expect(component.prop('current')).toBe(6)
    })

    it('should render pages', () => {
      const pages = component.find('PageItem')
      expect(pages).toHaveLength(11)
      expect(pages.at(0).prop('disabled')).toBeFalsy()
      expect(pages.at(1).prop('disabled')).toBeFalsy()
      expect(pages.at(2).prop('active')).toBeFalsy()
      expect(pages.at(3).prop('disabled')).toBeTruthy()
      expect(pages.at(4).prop('active')).toBeFalsy()
      expect(pages.at(5).prop('active')).toBeTruthy()
      expect(pages.at(6).prop('active')).toBeFalsy()
      expect(pages.at(7).prop('disabled')).toBeTruthy()
      expect(pages.at(8).prop('active')).toBeFalsy()
      expect(pages.at(9).prop('disabled')).toBeFalsy()
      expect(pages.at(10).prop('disabled')).toBeFalsy()
    })
  })

  describe('Last page', () => {
    beforeAll(() => {
      jest.resetAllMocks()

      component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)
      setCurrent.mockImplementation((newValue) => {
        component.setProps({ current: newValue })
      })

      act(() => {
        component.find('PageItem').at(8).find('a').simulate('click')
        component.update()
      })
    })

    it('should set current', () => {
      expect(setCurrent).toBeCalledTimes(1)
      expect(component.prop('current')).toBe(10)
    })

    it('should render pages', () => {
      const pages = component.find('PageItem')
      expect(pages).toHaveLength(8)
      expect(pages.at(0).prop('disabled')).toBeFalsy()
      expect(pages.at(1).prop('disabled')).toBeFalsy()
      expect(pages.at(2).prop('active')).toBeFalsy()
      expect(pages.at(3).prop('disabled')).toBeTruthy()
      expect(pages.at(4).prop('active')).toBeFalsy()
      expect(pages.at(5).prop('active')).toBeTruthy()
      expect(pages.at(6).prop('disabled')).toBeTruthy()
      expect(pages.at(7).prop('disabled')).toBeTruthy()
    })
  })
})
