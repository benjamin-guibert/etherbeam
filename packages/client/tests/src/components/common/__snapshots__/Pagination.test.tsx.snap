// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Ellipsed should match snapshot 1`] = `
<ul
  class="app-class pagination pagination-sm"
>
  <li
    class="page-item"
  >
    <a
      class="page-link"
      href="#"
      role="button"
    >
      <span
        aria-hidden="true"
      >
        «
      </span>
      <span
        class="sr-only"
      >
        First
      </span>
    </a>
  </li>
  <li
    class="page-item"
  >
    <a
      class="page-link"
      href="#"
      role="button"
    >
      <span
        aria-hidden="true"
      >
        ‹
      </span>
      <span
        class="sr-only"
      >
        Previous
      </span>
    </a>
  </li>
  <li
    class="page-item"
  >
    <a
      class="page-link"
      href="#"
      role="button"
    >
      1
    </a>
  </li>
  <li
    class="page-item disabled"
  >
    <span
      class="page-link"
      disabled=""
    >
      <span
        aria-hidden="true"
      >
        …
      </span>
      <span
        class="sr-only"
      >
        More
      </span>
    </span>
  </li>
  <li
    class="page-item"
  >
    <a
      class="page-link"
      href="#"
      role="button"
    >
      4
    </a>
  </li>
  <li
    class="page-item active"
  >
    <span
      class="page-link"
    >
      5
      <span
        class="sr-only"
      >
        (current)
      </span>
    </span>
  </li>
  <li
    class="page-item"
  >
    <a
      class="page-link"
      href="#"
      role="button"
    >
      6
    </a>
  </li>
  <li
    class="page-item disabled"
  >
    <span
      class="page-link"
      disabled=""
    >
      <span
        aria-hidden="true"
      >
        …
      </span>
      <span
        class="sr-only"
      >
        More
      </span>
    </span>
  </li>
  <li
    class="page-item"
  >
    <a
      class="page-link"
      href="#"
      role="button"
    >
      10
    </a>
  </li>
  <li
    class="page-item"
  >
    <a
      class="page-link"
      href="#"
      role="button"
    >
      <span
        aria-hidden="true"
      >
        ›
      </span>
      <span
        class="sr-only"
      >
        Next
      </span>
    </a>
  </li>
  <li
    class="page-item"
  >
    <a
      class="page-link"
      href="#"
      role="button"
    >
      <span
        aria-hidden="true"
      >
        »
      </span>
      <span
        class="sr-only"
      >
        Last
      </span>
    </a>
  </li>
</ul>
`;
